### ISSUE TITLE
P2 Repository lookup blocks UI

### ISSUE THREAD

**Issue by cleanerx** [bndtools/bnd#1951]
When doing a P2 query the UI blocks if the repository is rather large.
The Jobs API offers asynchronous execution that may update the UI once the content is received.

See the following stack:

```
"main" #1 prio=6 os_prio=0 tid=0x000000000277e800 nid=0x3b74 runnable [0x000000000014c000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at java.io.BufferedInputStream.read1(Unknown Source)
	at java.io.BufferedInputStream.read(Unknown Source)
	- locked <0x00000000f95ca528> (a java.io.BufferedInputStream)
	at sun.net.www.MeteredStream.read(Unknown Source)
	- locked <0x00000000ebb60e88> (a sun.net.www.http.KeepAliveStream)
	at java.io.FilterInputStream.read(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(Unknown Source)
	at aQute.bnd.http.ProgressWrappingStream.read(ProgressWrappingStream.java:69)
	at aQute.lib.io.IO.copy(IO.java:156)
	at aQute.lib.io.IO.copy(IO.java:146)
	at aQute.lib.io.IO.copy(IO.java:254)
	at aQute.bnd.http.URLCache$Info.update(URLCache.java:74)
	at aQute.bnd.http.HttpClient.doCached0(HttpClient.java:263)
	at aQute.bnd.http.HttpClient.doCached(HttpClient.java:154)
	at aQute.bnd.http.HttpClient.send(HttpClient.java:130)
	at aQute.bnd.http.HttpRequest$1.run(HttpRequest.java:209)
	at aQute.bnd.osgi.Processor$1.rejectedExecution(Processor.java:110)
	at java.util.concurrent.ThreadPoolExecutor.reject(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source)
	at aQute.bnd.http.HttpRequest.async(HttpRequest.java:204)
	at aQute.bnd.repository.p2.provider.P2Indexer.readRepository(P2Indexer.java:137)
	at aQute.bnd.repository.p2.provider.P2Indexer.<init>(P2Indexer.java:67)
	at aQute.bnd.repository.p2.provider.P2Repository.getP2Index0(P2Repository.java:70)
	at aQute.bnd.repository.p2.provider.P2Repository.getP2Index(P2Repository.java:53)
	- locked <0x00000000cd2347c0> (a aQute.bnd.repository.p2.provider.P2Repository)
	at aQute.bnd.repository.p2.provider.P2Repository.list(P2Repository.java:83)
	at bndtools.model.repo.RepositoryTreeContentProvider.getRepositoryBundles(RepositoryTreeContentProvider.java:250)
	at bndtools.model.repo.RepositoryTreeContentProvider.addCollection(RepositoryTreeContentProvider.java:178)
	at bndtools.model.repo.RepositoryTreeContentProvider.getElements(RepositoryTreeContentProvider.java:94)
	at org.eclipse.jface.viewers.StructuredViewer.getRawChildren(StructuredViewer.java:1000)
	at org.eclipse.jface.viewers.ColumnViewer.getRawChildren(ColumnViewer.java:700)
	at org.eclipse.jface.viewers.AbstractTreeViewer.getRawChildren(AbstractTreeViewer.java:1349)
	at org.eclipse.jface.viewers.TreeViewer.getRawChildren(TreeViewer.java:353)
	at org.eclipse.jface.viewers.StructuredViewer.getFilteredChildren(StructuredViewer.java:906)
	at org.eclipse.jface.viewers.AbstractTreeViewer.getSortedChildren(AbstractTreeViewer.java:617)
	at org.eclipse.jface.viewers.AbstractTreeViewer$1.run(AbstractTreeViewer.java:815)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
	at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:791)
	at org.eclipse.jface.viewers.TreeViewer.createChildren(TreeViewer.java:611)
	at org.eclipse.jface.viewers.AbstractTreeViewer.createChildren(AbstractTreeViewer.java:762)
	at org.eclipse.jface.viewers.AbstractTreeViewer.internalInitializeTree(AbstractTreeViewer.java:1541)
	at org.eclipse.jface.viewers.TreeViewer.internalInitializeTree(TreeViewer.java:790)
	at org.eclipse.jface.viewers.AbstractTreeViewer$5.run(AbstractTreeViewer.java:1525)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1462)
	at org.eclipse.jface.viewers.TreeViewer.preservingSelection(TreeViewer.java:366)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1423)
	at org.eclipse.jface.viewers.AbstractTreeViewer.inputChanged(AbstractTreeViewer.java:1517)
	at org.eclipse.jface.viewers.ContentViewer.setInput(ContentViewer.java:292)
	at org.eclipse.jface.viewers.StructuredViewer.setInput(StructuredViewer.java:1701)
	at bndtools.central.RepositoriesViewRefresher$1$1.run(RepositoriesViewRefresher.java:121)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:35)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:135)
	- locked <0x00000000cd5704e8> (a org.eclipse.swt.widgets.RunnableLock)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4155)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3772)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1127)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1018)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:694)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:606)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:139)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:669)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:608)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1515)

   Locked ownable synchronizers:
	- <0x00000000f9213ca0> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
```

**Comment by bjhargrave** [comment]
> at aQute.bnd.http.HttpRequest$1.run(HttpRequest.java:209) at aQute.bnd.osgi.Processor$1.rejectedExecution(Processor.java:110) at java.util.concurrent.ThreadPoolExecutor.reject(Unknown Source) at java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source) at aQute.bnd.http.HttpRequest.async(HttpRequest.java:204) at aQute.bnd.repository.p2.provider.P2Indexer.readRepository(P2Indexer.java:137) at aQute.bnd.repository.p2.provider.P2Indexer.<init>(P2Indexer.java:67)

I think this issue has the same root cause as #1747.

Both the P2 repo and the BndPom repo have to deeply traverse the dependencies from the root description (pom, p2 index). Both repo impls use executors to asynchronously divide the work and Promises to combine the result. 

The executor is part of the Processor class which supports 64 threads with a synchronous queue that runs rejected work on the calling thread. This means that when there are more than 64 current async work items, the 65th and later will steal the callers thread to run.

Under Bndtools, the caller's thread that gets stolen is the UI thread which can then lead to UI nonresponsiveness. So Bndtools needs to change to use a job to call RepositoryPlugin methods (like `list`) which can result in significant work being done by the repo impl as we see here for P2 and BndPom.

So basically, this is an issue that must be solved in Bndtools. It is calling a potentially long running repo operation on the Eclipse UI thread.

**Comment by gamerson** [comment]
Fixed in https://github.com/bndtools/bndtools/commit/ac4a86c5d9436ea6653cb15c44cde484d0308e1b
